From 9937c79febcfafc6883cda4d7263b05b16d24b40 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Mon, 18 Nov 2024 00:42:12 +0100
Subject: [PATCH] [DogeCash] Add `BlockManager` param to
 `CBlockIndex::GetBlockHeader` (#44)

This is needed later for loading the auxpow from disk.

(cherry picked from commit 46a77bf9e25f4bbe2ed832a6b0aef91772c436bb)

diff --git a/src/blockindex.cpp b/src/blockindex.cpp
index b5b4af8a8462d02a93e7bc57c82bc563bfd05b7d..59d2298ae9f722332a2fb9f23a482ac6572eb43c 100644
--- a/src/blockindex.cpp
+++ b/src/blockindex.cpp
@@ -4,6 +4,7 @@
 
 #include <blockindex.h>
 #include <logging.h>
+#include <node/blockstorage.h>
 #include <tinyformat.h>
 #include <util/check.h>
 
@@ -33,6 +34,27 @@ std::string CBlockIndex::ToString() const {
         nHeight, hashMerkleRoot.ToString(), GetBlockHash().ToString());
 }
 
+CBlockHeader
+CBlockIndex::GetBlockHeader(const node::BlockManager &blockman) const {
+    CBlockHeader block;
+    if (VersionHasAuxPow(nVersion)) {
+        if (!blockman.ReadBlockHeaderFromDisk(block, *this)) {
+            throw std::ios_base::failure(
+                "Failed reading AuxPow CBlockIndex header from disk");
+        }
+        return block;
+    }
+    block.nVersion = nVersion;
+    if (pprev) {
+        block.hashPrevBlock = pprev->GetBlockHash();
+    }
+    block.hashMerkleRoot = hashMerkleRoot;
+    block.nTime = nTime;
+    block.nBits = nBits;
+    block.nNonce = nNonce;
+    return block;
+}
+
 const CBlockIndex *CBlockIndex::GetAncestor(int height) const {
     if (height > nHeight || height < 0) {
         return nullptr;
diff --git a/src/blockindex.h b/src/blockindex.h
index e107855799141b6a4bb2c472f282c3ef17cd3c6f..be02e680f8076ce5fecbdbae66cb008c8c7ddbb8 100644
--- a/src/blockindex.h
+++ b/src/blockindex.h
@@ -16,6 +16,10 @@
 
 struct BlockHash;
 
+namespace node {
+class BlockManager;
+}
+
 /**
  * The block chain is a tree shaped structure starting with the genesis block at
  * the root, with each block potentially having multiple candidates to be the
@@ -113,18 +117,7 @@ public:
         return ret;
     }
 
-    CBlockHeader GetBlockHeader() const {
-        CBlockHeader block;
-        block.nVersion = nVersion;
-        if (pprev) {
-            block.hashPrevBlock = pprev->GetBlockHash();
-        }
-        block.hashMerkleRoot = hashMerkleRoot;
-        block.nTime = nTime;
-        block.nBits = nBits;
-        block.nNonce = nNonce;
-        return block;
-    }
+    CBlockHeader GetBlockHeader(const node::BlockManager &) const;
 
     BlockHash GetBlockHash() const {
         assert(phashBlock != nullptr);
diff --git a/src/headerssync.cpp b/src/headerssync.cpp
index d0d24f8db20e7b43c15b7bdf787ddb95c6addb6b..3baf50e894fb38e861066ee3911d943e539a750e 100644
--- a/src/headerssync.cpp
+++ b/src/headerssync.cpp
@@ -28,6 +28,7 @@ static_assert(sizeof(CompressedHeader) == 48);
 HeadersSyncState::HeadersSyncState(NodeId id,
                                    const Consensus::Params &consensus_params,
                                    const CBlockIndex *chain_start,
+                                   CBlockHeader chain_start_header,
                                    const arith_uint256 &minimum_required_work)
     : m_id(id), m_consensus_params(consensus_params),
       m_chain_start(chain_start),
@@ -35,7 +36,7 @@ HeadersSyncState::HeadersSyncState(NodeId id,
       m_current_chain_work(chain_start->nChainWork),
       m_commit_offset(
           FastRandomContext().randrange<unsigned>(HEADER_COMMITMENT_PERIOD)),
-      m_last_header_received(m_chain_start->GetBlockHeader()),
+      m_last_header_received(chain_start_header),
       m_current_height(chain_start->nHeight) {
     // Estimate the number of blocks that could possibly exist on the peer's
     // chain *right now* using 6 blocks/second (fastest blockrate given the MTP
diff --git a/src/headerssync.h b/src/headerssync.h
index ee3f081f17952bcb334c67bac93e8dfccd470d79..2f23dee5ab802cb3d6ee65ab64331adac22e2531 100644
--- a/src/headerssync.h
+++ b/src/headerssync.h
@@ -148,6 +148,7 @@ public:
      */
     HeadersSyncState(NodeId id, const Consensus::Params &consensus_params,
                      const CBlockIndex *chain_start,
+                     CBlockHeader chain_start_header,
                      const arith_uint256 &minimum_required_work);
 
     /** Result data structure for ProcessNextHeaders. */
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 8821e74cfc403d844329bc190a687f15dd35cd6c..7a5625bcea1e7163d23d7e30252e66568b6ca8d6 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -3971,9 +3971,10 @@ bool PeerManagerImpl::TryLowWorkHeadersSync(
             // of headers is known, some header in this set must be new, so
             // advancing to the first unknown header would be a small effect.
             LOCK(peer.m_headers_sync_mutex);
-            peer.m_headers_sync.reset(
-                new HeadersSyncState(peer.m_id, m_chainparams.GetConsensus(),
-                                     chain_start_header, minimum_chain_work));
+            peer.m_headers_sync.reset(new HeadersSyncState(
+                peer.m_id, m_chainparams.GetConsensus(), chain_start_header,
+                chain_start_header->GetBlockHeader(m_chainman.m_blockman),
+                minimum_chain_work));
 
             // Now a HeadersSyncState object for tracking this synchronization
             // is created, process the headers using it as normal. Failures are
@@ -5806,7 +5807,7 @@ void PeerManagerImpl::ProcessMessage(
                  hashStop.IsNull() ? "end" : hashStop.ToString(),
                  pfrom.GetId());
         for (; pindex; pindex = m_chainman.ActiveChain().Next(pindex)) {
-            vHeaders.push_back(pindex->GetBlockHeader());
+            vHeaders.push_back(pindex->GetBlockHeader(m_chainman.m_blockman));
             if (--nLimit <= 0 || pindex->GetBlockHash() == hashStop) {
                 break;
             }
@@ -8899,7 +8900,8 @@ bool PeerManagerImpl::SendMessages(const Config &config, CNode *pto) {
                     pBestIndex = pindex;
                     if (fFoundStartingHeader) {
                         // add this to the headers message
-                        vHeaders.push_back(pindex->GetBlockHeader());
+                        vHeaders.push_back(
+                            pindex->GetBlockHeader(m_chainman.m_blockman));
                     } else if (PeerHasHeader(&state, pindex)) {
                         // Keep looking for the first new block.
                         continue;
@@ -8908,7 +8910,8 @@ bool PeerManagerImpl::SendMessages(const Config &config, CNode *pto) {
                         // Peer doesn't have this header but they do have the
                         // prior one. Start sending headers.
                         fFoundStartingHeader = true;
-                        vHeaders.push_back(pindex->GetBlockHeader());
+                        vHeaders.push_back(
+                            pindex->GetBlockHeader(m_chainman.m_blockman));
                     } else {
                         // Peer doesn't have this header or the prior one --
                         // nothing will connect, so bail out.
diff --git a/src/rest.cpp b/src/rest.cpp
index b5f6e215ed3cbfcbb3ab14f8618d6c333664c7ee..93ecf851a5afc65921c28a8791d15bc06fb7af6c 100644
--- a/src/rest.cpp
+++ b/src/rest.cpp
@@ -254,7 +254,7 @@ static bool rest_headers(Config &config, const std::any &context,
         case RESTResponseFormat::BINARY: {
             DataStream ssHeader{};
             for (const CBlockIndex *pindex : headers) {
-                ssHeader << pindex->GetBlockHeader();
+                ssHeader << pindex->GetBlockHeader(GetChainman(context, req)->m_blockman);
             }
 
             std::string binaryHeader = ssHeader.str();
@@ -266,7 +266,7 @@ static bool rest_headers(Config &config, const std::any &context,
         case RESTResponseFormat::HEX: {
             DataStream ssHeader{};
             for (const CBlockIndex *pindex : headers) {
-                ssHeader << pindex->GetBlockHeader();
+                ssHeader << pindex->GetBlockHeader(GetChainman(context, req)->m_blockman);
             }
 
             std::string strHex = HexStr(ssHeader) + "\n";
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 2ced43c9f9f85b780662dcb67537551d65cd043c..901f3e05a6e1ef2db1fd2aaba152234e540897b1 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -643,20 +643,20 @@ static RPCHelpMan getblockheader() {
                 LOCK(cs_main);
                 pblockindex = chainman.m_blockman.LookupBlockIndex(hash);
                 tip = chainman.ActiveTip();
-            }
 
-            if (!pblockindex) {
-                throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY,
-                                   "Block not found");
-            }
+                if (!pblockindex) {
+                    throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY,
+                                       "Block not found");
+                }
 
             if (!fVerbose) {
                 DataStream ssBlock{};
-                ssBlock << pblockindex->GetBlockHeader();
+                ssBlock << pblockindex->GetBlockHeader(chainman.m_blockman);
                 std::string strHex = HexStr(ssBlock);
                 return strHex;
             }
 
+            }
             return blockheaderToJSON(*tip, *pblockindex);
         },
     };
diff --git a/src/test/blockindex_tests.cpp b/src/test/blockindex_tests.cpp
index 2e77cf33b368ecd7497cd5aeb119eb91c33c51c1..7903579213ab01255962bcaf0f0203a7cc267ea9 100644
--- a/src/test/blockindex_tests.cpp
+++ b/src/test/blockindex_tests.cpp
@@ -5,6 +5,7 @@
 #include <blockvalidity.h>
 #include <chain.h>
 #include <uint256.h>
+#include <validation.h>
 
 #include <test/util/setup_common.h>
 
@@ -30,7 +31,8 @@ BOOST_AUTO_TEST_CASE(get_block_header) {
 
     CBlockIndex index = CBlockIndex(header);
 
-    CBlockHeader checkHeader = index.GetBlockHeader();
+    CBlockHeader checkHeader =
+        index.GetBlockHeader(m_node.chainman->m_blockman);
     BOOST_CHECK(checkHeader.nVersion == expectedVersion);
     BOOST_CHECK(checkHeader.hashMerkleRoot == expectedMerkleRoot);
     BOOST_CHECK(checkHeader.nTime == expectedBlockTime);
diff --git a/src/test/headers_sync_chainwork_tests.cpp b/src/test/headers_sync_chainwork_tests.cpp
index 440a76c6c53dc3d132b04aa88a466088be0fa233..25179113c4360ef39a0d5746b69e29d273eabc2b 100644
--- a/src/test/headers_sync_chainwork_tests.cpp
+++ b/src/test/headers_sync_chainwork_tests.cpp
@@ -101,8 +101,9 @@ BOOST_AUTO_TEST_CASE(headers_sync_state) {
     headers_batch.insert(headers_batch.end(), std::next(first_chain.begin()),
                          first_chain.end());
 
-    hss.reset(new HeadersSyncState(0, Params().GetConsensus(), chain_start,
-                                   chain_work));
+    hss.reset(new HeadersSyncState(
+        0, Params().GetConsensus(), chain_start,
+        chain_start->GetBlockHeader(m_node.chainman->m_blockman), chain_work));
     (void)hss->ProcessNextHeaders({first_chain.front()}, true);
     // Pretend the first header is still "full", so we don't abort.
     auto result = hss->ProcessNextHeaders(headers_batch, true);
@@ -119,8 +120,9 @@ BOOST_AUTO_TEST_CASE(headers_sync_state) {
     BOOST_CHECK(hss->GetState() == HeadersSyncState::State::FINAL);
 
     // Now try again, this time feeding the first chain twice.
-    hss.reset(new HeadersSyncState(0, Params().GetConsensus(), chain_start,
-                                   chain_work));
+    hss.reset(new HeadersSyncState(
+        0, Params().GetConsensus(), chain_start,
+        chain_start->GetBlockHeader(m_node.chainman->m_blockman), chain_work));
     (void)hss->ProcessNextHeaders(first_chain, true);
     BOOST_CHECK(hss->GetState() == HeadersSyncState::State::REDOWNLOAD);
 
@@ -134,8 +136,9 @@ BOOST_AUTO_TEST_CASE(headers_sync_state) {
 
     // Finally, verify that just trying to process the second chain would not
     // succeed (too little work)
-    hss.reset(new HeadersSyncState(0, Params().GetConsensus(), chain_start,
-                                   chain_work));
+    hss.reset(new HeadersSyncState(
+        0, Params().GetConsensus(), chain_start,
+        chain_start->GetBlockHeader(m_node.chainman->m_blockman), chain_work));
     BOOST_CHECK(hss->GetState() == HeadersSyncState::State::PRESYNC);
     // Pretend just the first message is "full", so we don't abort.
     (void)hss->ProcessNextHeaders({second_chain.front()}, true);
